halo2_proofs/src/dev/cost_model.rs

```diff
@@ -0,0 +1,404 @@
+//! The cost estimator takes high-level parameters for a circuit design, and estimates the verification cost, as well as resulting proof size.
+
+use std::{
+    cmp, fmt, iter,
+    num::ParseIntError,
+    str::FromStr,
+    time::{Duration, Instant},
+};
+
+use crate::{arithmetic::best_multiexp, plonk::Circuit};
+use ff::{Field, FromUniformBytes};
+use group::{Curve, Group};
+use halo2curves::pasta::pallas;
+
+use super::MockProver;
+
+/// Structure storing scalars and bases used in multiexp estimation
+pub struct Estimator {
+    /// Scalars for estimating multiexp performance.
+    multiexp_scalars: Vec<pallas::Scalar>,
+    /// Bases for estimating multiexp performance.
+    multiexp_bases: Vec<pallas::Affine>,
+}
+
+impl fmt::Debug for Estimator {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Estimator")
+    }
+}
+
+impl Estimator {
+    fn random(k: usize) -> Self {
+        let max_size = 1 << (k + 1);
+        let mut rng = rand_core::OsRng;
+
+        Estimator {
+            multiexp_scalars: (0..max_size)
+                .map(|_| pallas::Scalar::random(&mut rng))
+                .collect(),
+            multiexp_bases: (0..max_size)
+                .map(|_| pallas::Point::random(&mut rng).to_affine())
+                .collect(),
+        }
+    }
+
+    fn multiexp(&self, size: usize) -> Duration {
+        let start = Instant::now();
+        best_multiexp(&self.multiexp_scalars[..size], &self.multiexp_bases[..size]);
+        Instant::now().duration_since(start)
+    }
```

Comment:
I don't like this approach TBH. 
The problem with this is that for large circuits this is the same as directly benchmarking the circuit, which is then not "estimating" anything and rather just running.
Also, any minimal change will require to re-run all the MSMs which is quite crazy. As again, for circuits >= `1<<21` it starts to be painful, specially considering there's no possible incremental compilation.

Reply:
Then we can maybe do this in another PR. But if you're up to fix this all within this PR, would be much better!